### ISSUE TITLE
New rule: Prefer logical assignment operators

### ISSUE THREAD

**Issue by glen-84** [eslint/eslint#13689]
**Please describe what the rule should do:**

Suggest the use of logical assignment operators.

**What new ECMAScript feature does this rule relate to?**

[Logical Assignment](https://github.com/tc39/proposal-logical-assignment/)

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[x] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

_Bad:_

```js
opts.foo = opts.foo ?? 'bar'
```

_Good:_

```js
opts.foo ??= 'bar'
```

(same for `||=` and `&&=`)

-- Optional extras --

_Bad:_

```js
if (!a) {
  a = 'foo';
}
```

_Good:_

```js
a ||= 'foo';
```

_Bad:_

```js
opts.baz ?? (opts.baz = 'qux');
```

_Good:_

```js
opts.baz ??= 'qux';
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It's a general syntax suggestion.

**Are you willing to submit a pull request to implement this rule?**

Not at this time.

**Comment by mdjermanovic** [comment]
Thanks for the new rule proposal! I'll champion this.

Our initial thoughts were to just add an option to [operator-assignment](https://eslint.org/docs/rules/operator-assignment), but it seems better to have a separate rule because it can cover more patterns, like `if (!a) { a = 'foo'; }`.

`operator-assignment` would additionally provide a way to disallow logical assignment operators (option `"never"`) if that's the preference, but that can be easily achieved with [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax).

**Comment by mdjermanovic** [comment]
> _Bad:_
> 
> ```js
> opts.foo = opts.foo ?? 'bar'
> ```
> 
> _Good:_
> 
> ```js
> opts.foo ??= 'bar'
> ```
> 
> (same for `||=` and `&&=`)

These are basic patterns that should be certainly supported for variables, but maybe we should add an option for member expressions, since these two expressions are not equivalent (observable if `foo` happens to be a setter).

At least, this shouldn't be auto-fixed if it isn't an assignment to a variable. Perhaps we could provide a [suggestion](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions) in that case.

**Comment by mdjermanovic** [comment]
> _Bad:_
> 
> ```js
> if (!a) {
>   a = 'foo';
> }
> ```
> 
> _Good:_
> 
> ```js
> a ||= 'foo';
> ```

Maybe we could also consider these:

```js
// a &&= 'foo'
if (a) a = 'foo';

// a ??= 'foo'
if (a == null) a = 'foo';

// a ??= 'foo'
if (a == undefined) a = 'foo';

// a ??= 'foo'
if (a === undefined || a === null) a = 'foo';
```

**Comment by DMartens** [comment]
What is the current status of this issue?
I offer to contribute this rule (I already have a personal rule with the basic patterns).
I think what is left to decide is:
1. Should additional patterns be supported?
-> I would say yes and would add patterns like `a = a !== undefined ? a : 'foo'`.
2. Create a new rule or add it to the existing rule?
-> I would prefer creating a new rule because of the possible additional patterns and this is then a non-breaking change.
3. What should the rule options be?
-> I would suggest the same as the `operator-assignment` rule (`always` or `never`) but maybe also disabling checking for the additional patterns.

**Comment by mdjermanovic** [comment]
> What is the current status of this issue?

It needs one more :+1: from a team member to be [accepted](https://eslint.org/docs/maintainer-guide/issues#accepting-issues).

> I offer to contribute this rule (I already have a personal rule with the basic patterns).

That will be great, thanks! 

Let's see first if the issue will be accepted and agree on the final spec.

> 1. Should additional patterns be supported?
>    -> I would say yes and would add patterns like `a = a !== undefined ? a : 'foo'`.

Make sense to support patterns that involve conditional expressions, though in some cases it might overlap with `no-unneeded-ternary`.

> 2. Create a new rule or add it to the existing rule?
>    -> I would prefer creating a new rule because of the possible additional patterns and this is then a non-breaking change.

I'd also prefer a new rule, for the same reasons.

> 3. What should the rule options be?
>    -> I would suggest the same as the `operator-assignment` rule (`always` or `never`) 

I think this rule shouldn't have `always`/`never`, because `never` doesn't quite fit the `prefer-` name of the rule. Also, it's easy to disallow logical assignment operators with the `no-restricted-syntax` rule.

> but maybe also disabling checking for the additional patterns.

Make sense to me, but would like more opinions on this. I'll try to make a single post with all patterns and options mentioned in this issue soon, so we could have a more concrete spec to evaluate.

**Comment by ljharb** [comment]
Using no-restricted-syntax isn't helpful because it can't autofix in the reverse direction.

In other words, it seems pretty useful to have a rule that can be configured to autofix from the long form to the logical operator form, or the inverse.

**Comment by nzakas** [comment]
@mdjermanovic whats the next steps on this?

**Comment by mdjermanovic** [comment]
Sorry for the delay, here's a detailed proposal.

Default configuration:

```json
{
    "logical-assignment-operators": ["error", "always", {
        "enforceForIfStatements": false
    }]
}
```

All examples are for "incorrect" code, `foo` can be any expression.

```js
/* eslint logical-assignment-operators: ["error", "always"] */

a = a || foo // autofix: a ||= foo
a = a && foo // autofix: a &&= foo
a = a ?? foo // autofix: a ??= foo

a || (a = foo) // autofix: a ||= foo
a && (a = foo) // autofix: a &&= foo
a ?? (a = foo) // autofix: a ??= foo

a.b = a.b || foo // suggested fix: a.b ||= foo
a.b = a.b && foo // suggested fix: a.b &&= foo
a.b = a.b ?? foo // suggested fix: a.b ??= foo

a.b || (a.b = foo) // autofix: a.b ||= foo
a.b && (a.b = foo) // autofix: a.b &&= foo
a.b ?? (a.b = foo) // autofix: a.b ??= foo
```

```js
/* eslint logical-assignment-operators: ["error", "always", { "enforceForIfStatements": true }] */

if (!a) a = foo; // autofix: a ||= foo;
if (a) a = foo; // autofix: a &&= foo;
if (a == undefined) a = foo; // autofix: a ??= foo;
if (a == null) a = foo; // autofix: a ??= foo;
if (a === undefined || a === null) a = foo; // autofix: a ??= foo;

if (!a.b) a.b = foo; // autofix: a.b ||= foo;
if (a.b) a.b = foo; // autofix: a.b &&= foo;
if (a.b == undefined) a.b = foo; // autofix: a ??= foo;
if (a.b == null) a.b = foo; // autofix: a ??= foo;
if (a.b === undefined || a.b === null) a.b = foo; // suggested fix: a ??= foo;
```

```js
/* eslint logical-assignment-operators: ["error", "never"] */

a ||= foo // autofix: a = a || foo
a &&= foo // autofix: a = a && foo
a ??= foo // autofix: a = a ?? foo

a.b ||= foo // suggested fix: a.b = a.b || foo
a.b &&= foo // suggested fix: a.b = a.b && foo
a.b ??= foo // suggested fix: a.b = a.b ?? foo
```

I didn't include conditional expressions because that would overlap with the `no-unneeded-ternary` rule (option [`defaultAssignment: false`](https://eslint.org/docs/rules/no-unneeded-ternary#defaultassignment)).

Notes:

* `enforceForIfStatements` option should not be available when the string option is set to `"never"`. We can enforce this in rule schema.
* "suggested fix" means that the rule provides a [suggestion](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions) instead of autofixing the code. The examples with "suggested fix" are those where the replacement code might not have the same behavior as the original code. In particular, because of possible getters/setters.
* Member expression can be more complex than `a.b`. We can use [`isSameReference`](https://github.com/eslint/eslint/blob/5891c7533f500110129fdea7b9b63c8a409da0bd/lib/rules/utils/ast-utils.js#L392) helper function. For example, `a.b.c = a.b.c || foo` should be reported, but `f(a).b = f(a).b || foo` should not.
* When the member expression is more complex, the fix must be a suggestion instead of autofix, because of possible getters. For example, `a.b.c || (a.b.c = foo)`. See [`canBeFixed`](https://github.com/eslint/eslint/blob/5891c7533f500110129fdea7b9b63c8a409da0bd/lib/rules/operator-assignment.js#L49) in the `operator-assignment` rule.
* `if` should not have `else`.
* All examples with `if` also apply to single-statement blocks. For example, `if (!a) { a = foo; }` should be reported.
* The left-right order in `if` conditions shouldn't matter. For example, `a === undefined || a === null` is the same as `null === a || undefined === a`.
* `void 0` can be treated the same as `undefined`. For example, `if (a == void 0) a = foo;` should be reported.
* Autofixes/suggestions should account for possible extra parentheses. For example, `(a) = ((a) || foo)`. It's okay to remove them, but the end result should be valid code (for example, the autofix/suggestion shouldn't remove a left parenthesis but leave the corresponding right parenthesis).
* If any comments exists in the range that is going to be replaced, we can skip providing the autofix/suggestion.

**Comment by ljharb** [comment]
It’s worth noting that the variable assignments would have the same getter risk inside a with statement in sloppy mode, or if the variables are globals (in any location)

**Comment by amitbeck** [comment]
@mdjermanovic Can we expect this rule to be released soon?

Also I suggest naming it `logical-operator-assignment` to align with [`operator-assignment`](https://eslint.org/docs/rules/operator-assignment).
> Default configuration:
> ```json
> {
>     "logical-assignment-operators": ["error", "always", {
>         "enforceForIfStatements": false
>    }]
> }
> ```

**Comment by nzakas** [comment]
@amitbeck no one is currently working on this rule.

**Comment by nzakas** [comment]
Thanks!
